content/blog/2022-04-07-compiling-black-with-mypyc-part-3.md

```diff
@@ -0,0 +1,185 @@
+---
+title: Compiling Black with mypyc, Pt. 3 - Deployment
+slug: compiling-black-with-mypyc-part-3
+description: &desc mypyc can compile Black and it provides respectable performance
+  gains, let's deploy it in production!
+summary: *desc
+tags: [black, mypyc]
+draft: true
+showToc: true
+date: 2022-04-07 12:02:00-04:00
+---
+
+This is part of the "*Compiling Black with mypyc*" series.
+
+- [Pt. 1 - Initial Steps](../compiling-black-with-mypyc-part-1/)
+- [Pt. 2 - Optimization](../compiling-black-with-mypyc-part-2/)
+- [Pt. 3 - Deployment](.) (you're reading this one)
+
+## Building compiled wheels with GitHub Actions
+
+With the mypyc branch functional and pretty fast, it was time to automate building the
+wheels. Not only does this make the release process easier, but **it also means I can
+easily build wheels for platforms I don't have access to**, like MacOS.
+
+To make the CI configuration process easier, I finally took a look at [cibuildwheel]. I
+won't go into my exact process since it was basically trial and error + my many dumb
+mistakes 😅 [^10] but here are some noteworthy takeaways:
+
+- If you're using `setuptools-scm`, you might have to do a full clone on CI so the tag
+  history is still available by build time
+
+- **Read cibuildwheel's documentation before writing any configuration!** Seriously, if I
+  did I wouldn't have to make [this commit][read-the-bloody-docs] adding `{project}` to
+  the test command
+
+- Set `CIBW_BUILD_VERBOSITY` to at least `1` because it will make debugging build errors
+  (and trust me you will get some!) so much nicer
+
+- If wheel sizes are an issue then passing `debug_level=0` to `mypyc.build.mypycify`
+  should help by stripping all debug information. Not great if you hit a bunch of
+  segfaults or similar though, so it's a tradeoff
+
+So yeah, 20-ish commits later I had a basic but functional setup which could compile
+wheels for x86-64 Windows, MacOS, and Linux from CPython 3.6 to CPython 3.10. Universial2
+and ARM variants were also supported for the shiny M1 platform. The workflow is a bit
+slow, but that's expected and not a big deal.
+
+If you're curious, you can find the workflow here: [ichard26/black-mypyc-wheels]
+
+> I don't recommend using my workflow to build your own workflows since it's a bit hacky
```

Comment:
```suggestion
> I don't recommend using my workflow to build your own wheels since it's a bit hacky
```

Reply:
Meh, my intention was "don't use my workflow as a reference when writing your own," but this makes more sense.